### PR TITLE
Emulate waitid(2) with kqueue(2) on OpenBSD

### PR DESCRIPTION
**Pull request by bsdkurt**

### CODE REVIEW

**Comment by bsdkurt** [comment]
Both FreeBSD and NetBSD have waitid(2) so this is purely OpenBSD only.

**Comment by krytarowski** [comment]
What's wrong in reaping exit value? Isn't it possible to just ignore it?

**Comment by bsdkurt** [comment]
Java_java_lang_ProcessHandleImpl_waitForProcessExit0() has two modes: 1) wait for child exit and reap exit status. 2) wait for process exit and not reap exit status. The second mode utilizes waitid() w/WNOWAIT which OpenBSD doesn’t have so I’m emulating it. Reaping exit status in mode 2 makes some of the regress tests fail.

**Comment by battleblow** [comment]
> Java_java_lang_ProcessHandleImpl_waitForProcessExit0() has two modes: 1) wait for child exit and reap exit status. 2) wait for process exit and not reap exit status. The second mode utilizes waitid() w/WNOWAIT which OpenBSD doesn’t have so I’m emulating it. Reaping exit status in mode 2 makes some of the regress tests fail.

Kurt, I'm interested in both the test results and which of the tests you're running on OpenBSD.

File: src/java.base/bsd/native/libjava/ProcessHandleImpl_bsd.c [hunk 1]
```diff
@@ -46,6 +46,17 @@
 #include <sys/user.h>  // For kinfo_proc
 #endif
 
+#if defined(__OpenBSD__)
+#include <sys/event.h> // For kqueue
+#include <sys/time.h>  // For kqueue
+#endif
+
+#define RESTARTABLE(_cmd, _result) do { \
```
Comment (battleblow) [inline src/java.base/bsd/native/libjava/ProcessHandleImpl_bsd.c:55]: This looks identical to the definition in src/java.base/unix/native/libjava/childproc.h.  I'd suggest just including that.

File: src/java.base/bsd/native/libjava/ProcessHandleImpl_bsd.c [hunk 2]
```diff
@@ -476,3 +487,44 @@ void os_getCmdlineAndUserInfo(JNIEnv *env, jobject jinfo, pid_t pid) {
...
 }
 
+#if defined(__OpenBSD__)
+int os_waitForProcessExitNoReap(pid_t pid) {
+    int kq, ret;
+    struct kevent evSet;
+    struct kevent event;
+
+    kq = kqueue();
+    if (kq < 0)
```
Comment (battleblow) [inline src/java.base/bsd/native/libjava/ProcessHandleImpl_bsd.c]: kq == -1 ?  No strong feelings though

**Comment by bsdkurt** [comment]
> > Java_java_lang_ProcessHandleImpl_waitForProcessExit0() has two modes: 1) wait for child exit and reap exit status. 2) wait for process exit and not reap exit status. The second mode utilizes waitid() w/WNOWAIT which OpenBSD doesn’t have so I’m emulating it. Reaping exit status in mode 2 makes some of the regress tests fail.
> 
> Kurt, I'm interested in both the test results and which of the tests you're running on OpenBSD.

These are the most direct test cases and how I am running them:
```
cd test/jdk
/home/truk/jdk/openjdk-jdk11u/build/bsd-x86_64-normal-server-fastdebug/images/jdk/bin/java \
  -jar /home/truk/jdk/jtreg/jtreg/lib/jtreg.jar \
  -Xmx512m -conc:3 \
  -verbose:fail \
  -w:./JTworkPH -r:./JTreportPH \
  -a -k:!headful \
  -jdk:/home/truk/jdk/openjdk-jdk11u/build/bsd-x86_64-normal-server-fastdebug/images/jdk \
  ./java/lang/ProcessHandle
```
When I was throwing UnsupportedOperationException for reapStatus == JNI_FALSE I had failures in multiple places throughout the tests/jdk results. I no longer have the results from this run so I can't give specifics as to the number. Upon review of those test case failures it became apparent that I needed to provide support for the no reap case. I tried a few quick solutions that were helpful but still problematic; e.g. always return 0 without waiting, always return NOT_A_CHILD without waiting and using waitpid(2) which reaps. All of these had issues with test cases outside of just java/lang/ProcessHandle.

With the current implementation with emulated waitid I have some remaining failures in the java/langProcessHandle tests but they are tangential to this correction. As far as I can tell I have no failures in the full test suite due to the emulated waitid I have used.

* test InfoTest.test3(): failure
This test case is expecting a full path to the executable while I am only providing arg[0] string which can be relative or resolved via PATH.
* test OnExitTest.peerOnExitTest(): failure and test OnExitTest.test2(): failure
Test calls ProcessHandle.info() on a zombie process which raises java.lang.RuntimeException: No such process

My emulated waitid is not perfect but works well enough for the test suite and I think its the best I can do short of having waitid w/WNOWAIT. With waitid & WNOWAIT you can get the exit status of a zombie process without reaping the zombie, including if the process was terminated by a signal. waitid & WNOWAIT can be called multiple times on the same pid without side effects. With my kqueue implementation I can only get the exit status (without signal flags) if the event is placed before the process exits. Otherwise, if the event is placed on a zombie or non-existent process I have to return a value to let the caller know the process is gone and 0 was the best fit in this case.

File: src/java.base/bsd/native/libjava/ProcessHandleImpl_bsd.c (hunk 1 above)
Comment (bsdkurt) [reply src/java.base/bsd/native/libjava/ProcessHandleImpl_bsd.c:55]: > This looks identical to the definition in src/java.base/unix/native/libjava/childproc.h. I'd suggest just including that.

I'd rather not include an unrelated header to bring in this define. childproc.h is a header that is tightly coupled with childproc.c, not a common include header. It has a `/* TODO: Refactor. */` note there as well indicating the copying of the RESTARTABLE macro is known need refactoring into a common header. I'd prefer to let oracle do the refactoring at some future time and remove the local copy then.

**Comment by bsdkurt** [comment]
Are there any further comments or changes requested? I'll squash the commits together when I merge it in.

**Comment by battleblow** [comment]
> Are there any further comments or changes requested? I'll squash the commits together when I merge it in.

Please go ahead and merge it.

**Pull request merged**